File: backend/src/main/java/sn/immosn/backend/contrat/service/ContratService.java
package sn.immosn.backend.contrat.service;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.web.multipart.MultipartFile;
import sn.immosn.backend.client.web.contrat.dto.*;
import sn.immosn.backend.contrat.data.entity.StatutContrat;
import sn.immosn.backend.contrat.data.entity.TypeContrat;
import sn.immosn.backend.visite.data.entity.DemandeVisite;

public interface ContratService {

    ContratResponseDto create(ContratCreateRequestDto request);

    /**
     * Crée automatiquement un contrat depuis une visite clôturée.
     * Pré-remplit client, annonce, lead et montant (annonce.prix) depuis la visite.
     * L'admin ne fournit que typeContrat et, si LOCATION, dureeLocationMois.
     */
    ContratResponseDto createFromVisite(DemandeVisite visite, TypeContrat typeContrat, Integer dureeLocationMois);

    Page<ContratResponseDto> getClientContrats(String clientEmail, StatutContrat statut, Pageable pageable);

    Page<ContratResponseDto> getAllContrats(StatutContrat statut, Pageable pageable);

    ContratResponseDto getById(Long id, String userEmail, boolean isAdmin);

    /**
     * Met à jour le contrat. Si request.statut() == ACTIF, seul un SUPER_ADMIN peut effectuer
     * cette transition (vérifié via callerEmail) — déclenche aussi la création du compte client
     * depuis le prospect si nécessaire.
     */
    ContratResponseDto update(Long id, ContratUpdateRequestDto request, String callerEmail);

    /**
     * Met à jour les informations du prospect lié au contrat (nom, prénom, email, téléphone).
     * Permet à l'admin de corriger/compléter les infos avant l'activation, qui déclenchera
     * la création automatique du compte client à partir de ces données.
     * Échoue si le contrat n'a pas de prospect (déjà lié à un client).
     */
    ContratResponseDto updateProspect(Long id, ContratProspectUpdateRequestDto request);

    ContratResponseDto uploadDocument(Long id, MultipartFile file);

    /**
     * Ajoute une ou plusieurs pièces jointes typées au contrat (pré-contrat, pièce d'identité, etc.).
     * Les fichiers sont déjà uploadés (Cloudinary côté frontend) — seules les URLs résultantes
     * sont enregistrées ici, associées à leur type et à l'admin qui les a ajoutées.
     */
    ContratResponseDto addDocuments(Long id, ContratDocumentCreateRequestDto request, String adminEmail);

    ContratResponseDto removeDocument(Long id, Long documentId);

    ContratResponseDto demanderResiliation(Long id, ContratActionDto dto, String clientEmail);

    ContratResponseDto demanderProlongation(Long id, ContratActionDto dto, String clientEmail);

    ContratResponseDto accepterResiliation(Long id, ContratActionDto dto);

    ContratResponseDto refuserResiliation(Long id, ContratActionDto dto);

    ContratResponseDto accepterProlongation(Long id, ContratActionDto dto);

    ContratResponseDto refuserProlongation(Long id, ContratActionDto dto);
}

File: frontend/src/services/contratService.js
import api from './api'

export default {
  // POST /api/v1/contrats — ADMIN
  create(data) {
    return api.post('/contrats', data)
  },

  // GET /api/v1/contrats/client?page&size&statut
  getClientContrats(page = 0, size = 10, statut = null) {
    const params = { page, size }
    if (statut) params.statut = statut
    return api.get('/contrats/client', { params })
  },

  // GET /api/v1/contrats/admin?page&size&statut&typeContrat
  getAllContrats(page = 0, size = 20, statut = null, typeContrat = null) {
    const params = { page, size }
    if (statut) params.statut = statut
    if (typeContrat) params.typeContrat = typeContrat
    return api.get('/contrats/admin', { params })
  },

  // GET /api/v1/contrats/{id}
  getById(id) {
    return api.get(`/contrats/${id}`)
  },

  // PUT /api/v1/contrats/{id} — ADMIN
  update(id, data) {
    return api.put(`/contrats/${id}`, data)
  },

  // PUT /api/v1/contrats/{id}/prospect — ADMIN — { nom, prenom, email, telephone }
  updateProspect(id, data) {
    return api.put(`/contrats/${id}/prospect`, data)
  },

  // POST /api/v1/contrats/{id}/document — ADMIN (multipart)
  uploadDocument(id, file) {
    const form = new FormData()
    form.append('file', file)
    return api.post(`/contrats/${id}/document`, form, {
      headers: { 'Content-Type': 'multipart/form-data' },
    })
  },

  // PUT /api/v1/contrats/{id}/resiliation — CLIENT
  demanderResiliation(id, motif) {
    return api.put(`/contrats/${id}/resiliation`, { motif })
  },

  // PUT /api/v1/contrats/{id}/prolongation — CLIENT
  demanderProlongation(id, nouvelleDate, motif) {
    return api.put(`/contrats/${id}/prolongation`, { nouvelleDate, motif })
  },

  // PUT /api/v1/contrats/{id}/resiliation/accepter — ADMIN
  accepterResiliation(id, motif = null) {
    return api.put(`/contrats/${id}/resiliation/accepter`, motif ? { motif } : {})
  },

  // PUT /api/v1/contrats/{id}/resiliation/refuser — ADMIN
  refuserResiliation(id, motif = null) {
    return api.put(`/contrats/${id}/resiliation/refuser`, motif ? { motif } : {})
  },

  // PUT /api/v1/contrats/{id}/prolongation/accepter — ADMIN
  // nouvelleDate : nouvelle date de fin (LocalDate ISO)
  accepterProlongation(id, nouvelleDate = null, motif = null) {
    return api.put(`/contrats/${id}/prolongation/accepter`, { nouvelleDate, motif })
  },

  // PUT /api/v1/contrats/{id}/prolongation/refuser — ADMIN
  refuserProlongation(id, motif = null) {
    return api.put(`/contrats/${id}/prolongation/refuser`, motif ? { motif } : {})
  },

  // ─── Pièces jointes typées (pré-contrat, contrat signé, pièce d'identité, etc.) ───

  // POST /api/v1/contrats/{id}/documents — ADMIN — documents : [{ type, url }]
  addDocuments(id, documents) {
    return api.post(`/contrats/${id}/documents`, { documents })
  },

  // DELETE /api/v1/contrats/{id}/documents/{documentId} — ADMIN
  removeDocument(id, documentId) {
    return api.delete(`/contrats/${id}/documents/${documentId}`)
  },
}
